src/openms/source/ANALYSIS/DECHARGING/MetaboliteFeatureDeconvolution.cpp

```diff
@@ -0,0 +1,1209 @@
+    cons_map_p = ConsensusMap();
+
+    Int q_min = param_.getValue("charge_min");
+    Int q_max = param_.getValue("charge_max");
+    Int q_span = param_.getValue("charge_span_max");
+    Size max_neutrals = param_.getValue("max_neutrals");
+
+    double rt_diff_max = param_.getValue("retention_max_diff");
+    double rt_diff_max_local = param_.getValue("retention_max_diff_local");
+
+    double mz_diff_max = param_.getValue("mass_max_diff");
+
+    double rt_min_overlap = param_.getValue("min_rt_overlap");
+
+
+    // sort by RT and then m/z
+    fm_out = fm_in;
+    fm_out.sortByPosition();
+    fm_out.applyMemberFunction(&UniqueIdInterface::ensureUniqueId);
+    FeatureMapType fm_out_untouched = fm_out;
+
+
+    // search for most & least probable adduct to fix p threshold
+    double adduct_lowest_log_p = log(1.0);
+    double adduct_highest_log_p = log(0.0000000001);
+    for (Size i = 0; i < potential_adducts_.size(); ++i)
+    {
+      adduct_lowest_log_p  = std::min(adduct_lowest_log_p, potential_adducts_[i].getLogProb());
+      adduct_highest_log_p = std::max(adduct_highest_log_p, potential_adducts_[i].getLogProb());
+    }
+    Int max_minority_bound = param_.getValue("max_minority_bound");
+    double thresh_logp = adduct_lowest_log_p * max_minority_bound +
+                         adduct_highest_log_p * std::max(q_max - max_minority_bound, 0);
+
+
+    // create mass difference list
+
+    LOG_INFO << "Generating Masses with threshold: " << thresh_logp << " ...\n";
+    MassExplainer me(potential_adducts_, q_min, q_max, q_span, thresh_logp, max_neutrals);
+    me.compute();
+    LOG_INFO << "done\n";
+
+    // holds query results for a mass difference
+    MassExplainer::CompomerIterator md_s, md_e;
+    Compomer null_compomer(0, 0, -std::numeric_limits<double>::max());
+    SignedSize hits(0);
+
+    CoordinateType mz1, mz2, m1;
+
+    Size possibleEdges(0), overallHits(0);
+
+    // edges
+    PairsType feature_relation;
+    // for each feature, hold the explicit adduct type induced by edges
+    Map<Size, std::set<CmpInfo_> > feature_adducts;
+
+    // # compomer results that either passed or failed the feature charge constraints
+    Size no_cmp_hit(0), cmp_hit(0);
+
+    /*DoubleList dl_massdiff;
+    IntList il_chargediff;*/
+
+    // Backbone adduct: implicit adducts don't cost anything
+    Adduct proton(1, 1, Constants::PROTON_MASS_U, "H1", log(1.0), 0);
+
+    for (Size i_RT = 0; i_RT < fm_out.size(); ++i_RT) // ** RT-sweep line
+    {
+      mz1 = fm_out[i_RT].getMZ();
+
+      for (Size i_RT_window = i_RT + 1
+           ; (i_RT_window < fm_out.size())
+          && ((fm_out[i_RT_window].getRT() - fm_out[i_RT].getRT()) <= rt_diff_max)
+           ; ++i_RT_window)
+      { // ** RT-window
+
+        // knock-out criterion first: RT overlap
+        // use sorted structure and use 2nd start--1stend / 1st start--2ndend
+        const Feature& f1 = fm_out[i_RT];
+        const Feature& f2 = fm_out[i_RT_window];
```

Comment:
how would this work with a negative RT shift, e.g. if the adduct causes it to shift back in RT ?